### PR TITLE
feat(gamelift): add GameServerGroup L2 Construct for GameLift

### PR DESCRIPTION
**Pull request by stevehouel**
Following https://github.com/aws/aws-cdk-rfcs/issues/436 I have written the Gamelift FleetIQ L2 resource which create a GameServerGroup resource.

----

### All Submissions:

* [x] Have you followed the guidelines in our [Contributing guide?](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md)

### Adding new Unconventional Dependencies:

* [ ] This PR adds new unconventional dependencies following the process described [here](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md/#adding-new-unconventional-dependencies)

### New Features

* [x] Have you added the new feature to an [integration test](https://github.com/aws/aws-cdk/blob/main/INTEGRATION_TESTS.md)?
	* [x] Did you use `yarn integ` to deploy the infrastructure and generate the snapshot (i.e. `yarn integ` without `--dry-run`)?

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk/aws-gamelift/lib/game-server-group.ts [hunk 1]
```diff
@@ -0,0 +1,563 @@
...
+      weight: weight,
+    });
+  }
+
+  /**
+   * Adds an instance definition that GameLift FleetIQ can use when balancing and automatically scaling instances in the corresponding Auto Scaling group
+   *
+   * @param instanceDefinition The instance definition to add
+   */
+  public addInternalInstanceDefinition(instanceDefinition: InstanceDefinition) {
```
Comment (otaviomacedo) [inline packages/@aws-cdk/aws-gamelift/lib/game-server-group.ts:524]: What's the difference between these two methods? Should this one be private?

Also, what's the use case for this method? In which scenario would it not be enough to provide all instance definitions in the constructor props?

File: packages/@aws-cdk/aws-gamelift/lib/game-server-group.ts [hunk 2]
```diff
@@ -0,0 +1,563 @@
...
+   * A list of labels to assign to the new game server group resource.
+   * Tags are developer-defined key-value pairs.
+   * Tagging AWS resources is useful for resource management, access management, and cost allocation.
+   * Once the resource is created, you can use `TagResource`, `UntagResource`, and `ListTagsForResource` to add, remove, and view tags, respectively.
+   *
+   * @see https://docs.aws.amazon.com/general/latest/gr/aws_tagging.html
+   *
+   * @default {} - no tags
+   */
+  readonly tags?: { [key: string]: string };
```
Comment (otaviomacedo) [inline packages/@aws-cdk/aws-gamelift/lib/game-server-group.ts:334]: Where is this used?

File: packages/@aws-cdk/aws-gamelift/lib/game-server-group.ts [hunk 3]
```diff
@@ -0,0 +1,563 @@
...
+
+    const gameServerGroupArn = attrs.gameServerGroupArn ?? cdk.Stack.of(scope).formatArn({
+      service: 'gamelift',
+      resource: 'gameservergroup',
+      resourceName: attrs.gameServerGroupName,
+      arnFormat: cdk.ArnFormat.SLASH_RESOURCE_NAME,
+    });
+
+    class Import extends GameServerGroupBase {
+      public readonly gameServerGroupName = gameServerGroupName!;
```
Comment (otaviomacedo) [inline packages/@aws-cdk/aws-gamelift/lib/game-server-group.ts:382]: Do we need the `!` here?

File: packages/@aws-cdk/aws-gamelift/lib/game-server-group.ts [hunk 4]
```diff
@@ -0,0 +1,563 @@
...
+    const resource = new CfnGameServerGroup(this, 'Resource', {
+      gameServerGroupName: this.physicalName,
+      autoScalingPolicy: this.parseAutoScalingPolicy(props),
+      deleteOption: props.deleteOption,
+      balancingStrategy: props.balancingStrategy,
+      gameServerProtectionPolicy: props.protectGameServer ? 'FULL_PROTECTION' : 'NO_PROTECTION',
+      instanceDefinitions: cdk.Lazy.any({ produce: () => this.parseInstanceDefinitions() }),
+      launchTemplate: this.parseLaunchTemplate(props),
+      minSize: props.minSize,
+      maxSize: props.maxSize,
```
Comment (otaviomacedo) [inline packages/@aws-cdk/aws-gamelift/lib/game-server-group.ts:485]: These two properties have a minimum value (0 and 1, respectively). We should add a validation for this.

File: packages/@aws-cdk/aws-gamelift/lib/game-server-group.ts (hunk 1 above)
Comment (stevehouel) [reply packages/@aws-cdk/aws-gamelift/lib/game-server-group.ts:524]: Agreed. No real use cases for using this methods outside of the constructor

File: packages/@aws-cdk/aws-gamelift/lib/game-server-group.ts (hunk 4 above)
Comment (stevehouel) [reply packages/@aws-cdk/aws-gamelift/lib/game-server-group.ts:485]: ok will do

File: packages/@aws-cdk/aws-gamelift/lib/game-server-group.ts (hunk 3 above)
Comment (stevehouel) [reply packages/@aws-cdk/aws-gamelift/lib/game-server-group.ts:382]: Yes, as we are either extracting this value from ARN or getting this value from Attributes. and if value is null we are throwing an error but initial variable instance is String | undefined which create a compiler error

**Review by otaviomacedo — APPROVED** [review]

**Pull request merged**